### PR TITLE
NULL argument checks

### LINKED ISSUE #1673: NULL Checks for output parameters
**Originally by _dinan_ on 2012-08-01 09:58:53 -0500**

---

Many functions are missing NULL checks, e.g.:

MPIR_ERRTEST_ARGNULL(newcomm, "newcomm", mpi_errno);

for output parameters.  This can lead to bugs in the user program, which cause MPICH to seg. fault when dereferencing these bad pointers.  The attached test demonstrates that this for a communicator creation routine.

### PR DESCRIPTION
**Pull request by Bayyapu**
When the User makes a call to MPI_Comm_dup(MPI_COMM_WORLD, NULL) the program gives Segmentation error. To avoid this and capture the error, the newcomm argument is checked for NULL in comm_dup.c.  
This PR is to fix #1673

### CODE REVIEW

File: src/mpi/comm/comm_dup.c [hunk 1]
```diff
@@ -72,17 +72,13 @@ int MPIR_Comm_dup_impl(MPIR_Comm *comm_ptr, MPIR_Comm **newcomm_ptr)
 #undef FCNAME
 #define FCNAME MPL_QUOTE(FUNCNAME)
 /*@
-
```
Comment (raffenet) [inline src/mpi/comm/comm_dup.c]: whitespace changes here are not related to the functionality change, so they shouldn't be included. see the "what makes a good commit" section here: https://wiki.mpich.org/mpich/index.php/Version_Control_Systems_101

File: src/mpi/comm/comm_dup.c [hunk 2]
```diff
@@ -149,8 +140,16 @@ int MPI_Comm_dup(MPI_Comm comm, MPI_Comm *newcomm)
             /* Validate comm_ptr */
             MPIR_Comm_valid_ptr( comm_ptr, mpi_errno, FALSE );
             if (mpi_errno) goto fn_fail;
+            /*If newcomm is NULL, creates and error to exit*/
+	    if(newcomm==NULL)
```
Comment (raffenet) [inline src/mpi/comm/comm_dup.c]: The `MPIR_ERRTEST_ARGNULL` macro below should already be doing this. I think the case of `MPI_Comm_dup`, this check is already present.

**Comment by Bayyapu** [comment]
Few MPI functions that do not have a NULL check are being added to the new commit of PR #1673

File: src/mpi/datatype/type_get_contents.c [hunk 3]
```diff
@@ -122,6 +122,9 @@ int MPI_Type_get_contents(MPI_Datatype datatype,
             MPIR_Datatype_valid_ptr(datatype_ptr, mpi_errno);
 	    /* If comm_ptr is not value, it will be reset to null */
             if (mpi_errno) goto fn_fail;
+            MPIR_ERRTEST_ARGNULL(array_of_integers, "aointegers", mpi_errno);
+            MPIR_ERRTEST_ARGNULL(array_of_addressess, "aoaddr", mpi_errno);
```
Comment (raffenet) [inline src/mpi/datatype/type_get_contents.c]: Typo in variable name causes a compile error here.

File: src/mpi/comm/comm_split.c [hunk 4]
```diff
@@ -477,10 +478,8 @@ int MPI_Comm_split(MPI_Comm comm, int color, int key, MPI_Comm *newcomm)
     
     mpi_errno = MPIR_Comm_split_impl(comm_ptr, color, key, &newcomm_ptr);
     if (mpi_errno) MPIR_ERR_POP(mpi_errno);
-    if (newcomm_ptr)
```
Comment (raffenet) [inline src/mpi/comm/comm_split.c]: This change also looks unrelated to the commit it is in, but I also think it is incorrect. `MPIR_Comm_split_impl` will set `newcomm_ptr` to `NULL` in the case where a process supplies `MPI_UNDEFINED` in the color argument. The user code will then get `MPI_COMM_NULL` as the returned handle.

File: src/mpi/comm/intercomm_create.c [hunk 5]
```diff
@@ -261,6 +261,7 @@ int MPI_Intercomm_create(MPI_Comm local_comm, int local_leader,
 		    MPIR_ERRTEST_COMM(peer_comm, mpi_errno);
 		}
 	    }
+              MPIR_ERRTEST_ARGNULL(newintercomm, "newintercomm", mpi_errno);
```
Comment (raffenet) [inline src/mpi/comm/intercomm_create.c]: Indentation looks off here.

File: src/mpi/comm/intercomm_create.c [hunk 6]
```diff
@@ -303,6 +304,7 @@ int MPI_Intercomm_create(MPI_Comm local_comm, int local_leader,
 		    MPIR_ERR_SET(mpi_errno,MPI_ERR_RANK,"**ranksdistinct");
 		}
 		if (mpi_errno) goto fn_fail;
+                MPIR_ERRTEST_ARGNULL(newintercomm, "newintercomm", mpi_errno);
```
Comment (raffenet) [inline src/mpi/comm/intercomm_create.c]: Indentation off here.

File: src/mpi/datatype/type_get_envelope.c [hunk 7]
```diff
@@ -124,6 +124,10 @@ int MPI_Type_get_envelope(MPI_Datatype datatype,
             MPIR_Datatype_valid_ptr(datatype_ptr, mpi_errno);
 	    /* If comm_ptr is not value, it will be reset to null */
             if (mpi_errno != MPI_SUCCESS) goto fn_fail;
+	    MPIR_ERRTEST_ARGNULL(num_integers, "num_integers", mpi_errno);
```
Comment (raffenet) [inline src/mpi/datatype/type_get_envelope.c]: indentation

File: src/mpi/datatype/type_vector.c [hunk 8]
```diff
@@ -126,6 +126,7 @@ int MPI_Type_vector(int count,
 		MPIR_Datatype_valid_ptr(old_ptr, mpi_errno);
                 if (mpi_errno) goto fn_fail;
 	    }
+            MPIR_ERRTEST_ARGNULL(newtype, "newtype", mpi_errno);
```
Comment (raffenet) [inline src/mpi/datatype/type_vector.c]: indentation

File: src/mpi/group/group_range_excl.c [hunk 9]
```diff
@@ -192,6 +192,7 @@ int MPI_Group_range_excl(MPI_Group group, int n, int ranges[][3],
 							   ranges, n );
 	    }
             if (mpi_errno) goto fn_fail;
+	    MPIR_ERRTEST_ARGNULL(newgroup, "newgroup", mpi_errno);
```
Comment (raffenet) [inline src/mpi/group/group_range_excl.c]: indent

File: src/mpi/group/group_range_incl.c [hunk 10]
```diff
@@ -168,6 +168,7 @@ int MPI_Group_range_incl(MPI_Group group, int n, int ranges[][3],
 							   ranges, n );
 	    }
             if (mpi_errno) goto fn_fail;
+	    MPIR_ERRTEST_ARGNULL(newgroup, "newgroup", mpi_errno);
```
Comment (raffenet) [inline src/mpi/group/group_range_incl.c]: indent

File: src/mpi/spawn/comm_join.c [hunk 11]
```diff
@@ -180,7 +180,7 @@ int MPI_Comm_join(int fd, MPI_Comm *intercomm)
 
     mpi_errno = MPIR_Close_port_impl(local_port);
     if (mpi_errno) MPIR_ERR_POP(mpi_errno);
-
+    MPIR_ERRTEST_ARGNULL(intercomm, "intercomm", mpi_errno);
```
Comment (raffenet) [inline src/mpi/spawn/comm_join.c]: This also removes a line of whitespace. We should try not to do that when doing a simple addition, such as this.

File: src/mpi/spawn/comm_spawn.c [hunk 12]
```diff
@@ -107,6 +107,8 @@ int MPI_Comm_spawn(const char *command, char *argv[], int maxprocs, MPI_Info inf
 		MPIR_ERRTEST_ARGNULL(command, "command", mpi_errno);
 		MPIR_ERRTEST_ARGNEG(maxprocs, "maxprocs", mpi_errno);
 	    }
+            MPIR_ERRTEST_ARGNULL(intercomm, "intercomm", mpi_errno);
```
Comment (raffenet) [inline src/mpi/spawn/comm_spawn.c]: indent

File: src/mpi/spawn/comm_spawn_multiple.c [hunk 13]
```diff
@@ -119,6 +119,8 @@ int MPI_Comm_spawn_multiple(int count, char *array_of_commands[],
 		    MPIR_ERRTEST_ARGNEG(array_of_maxprocs[i], "array_of_maxprocs[i]", mpi_errno);
 		}
 	    }
+            MPIR_ERRTEST_ARGNULL(intercomm, "intercomm", mpi_errno);
```
Comment (raffenet) [inline src/mpi/spawn/comm_spawn_multiple.c]: indent

File: src/mpi/topo/cart_coords.c [hunk 14]
```diff
@@ -92,6 +92,7 @@ int MPI_Cart_coords(MPI_Comm comm, int rank, int maxdims, int coords[])
 	    if (mpi_errno != MPI_SUCCESS) goto fn_fail;
 
 	    MPIR_ERRTEST_RANK(comm_ptr, rank, mpi_errno);
+            MPIR_ERRTEST_ARGNULL(coords, "coords", mpi_errno);
```
Comment (raffenet) [inline src/mpi/topo/cart_coords.c]: indent

**Comment by raffenet** [comment]
test:jenkins/ch3/tcp

File: src/mpi/comm/comm_dup.c [hunk 15]
```diff
@@ -178,7 +178,6 @@ int MPI_Comm_dup(MPI_Comm comm, MPI_Comm *newcomm)
 	    "**mpi_comm_dup %C %p", comm, newcomm);
     }
 #   endif
-    *newcomm = MPI_COMM_NULL;
```
Comment (raffenet) [inline src/mpi/comm/comm_dup.c]: The only problem with this is that we still want to return `MPI_COMM_NULL` in the case where comm creation failed, and the user provided a valid comm handle ptr. Can you modify the patch to instead protect this assignment? Something like:
```
if (newcomm)
    *newcomm = MPI_COMM_NULL;
```

File: src/mpi/comm/comm_dup.c [hunk 16]
```diff
@@ -160,8 +160,10 @@ int MPI_Comm_dup(MPI_Comm comm, MPI_Comm *newcomm)
     
     mpi_errno = MPIR_Comm_dup_impl(comm_ptr, &newcomm_ptr);
     if (mpi_errno) MPIR_ERR_POP(mpi_errno);
-
-    MPIR_OBJ_PUBLISH_HANDLE(*newcomm, newcomm_ptr->handle);
+    if (newcomm_ptr)
```
Comment (raffenet) [inline src/mpi/comm/comm_dup.c]: This isn't quite right. If we are at this point in the code, we know that we have been given a valid pointer in `newcomm`. It should be safe to unconditionally publish the comm handle here. We only return `MPI_COMM_NULL` when the operation fails, which should be handled by the code after the `fn_fail` label.

**Comment by raffenet** [comment]
@Bayyapu this looks good! I am ready to sign off on both these commits after I trigger the tests to make sure everything works as expected.
test:jenkins/ch3/tcp

File: test/mpi/errors/comm/comm_create_group_test.c [hunk 17]
```diff
@@ -0,0 +1,31 @@
...
+    MPI_Comm_size(MPI_COMM_WORLD, &size);
+
+    MPI_Comm_dup(MPI_COMM_WORLD, &dup_comm);
+    MPI_Comm_group(dup_comm, &group);
+    MPI_Errhandler_set(MPI_COMM_WORLD, MPI_ERRORS_RETURN);
+
+    /*test comm_create_group for NULL variable */
+    mpi_errno = MPI_Comm_create_group(dup_comm, group, 10, NULL);
+    MPI_Error_class(mpi_errno, &errclass);
+    if(errclass == MPI_ERR_ARG)
```
Comment (raffenet) [inline test/mpi/errors/comm/comm_create_group_test.c]: As we discussed in person, this comparison should be `if (errclass != MPI_ERR_ARG)`. Also, make sure there is a space between the `if` and open parenthesis in your patches.

**Comment by raffenet** [comment]
Everything looks good to me now! Going to trigger the tests one more time.
test:jenkins/ch3/tcp

**Comment by raffenet** [comment]
test:jenkins/ch3/tcp

**Comment by raffenet** [comment]
test:jenkins/ch3/tcp

File: test/mpi/errors/comm/Makefile.am [hunk 18]
```diff
@@ -12,5 +12,6 @@ EXTRA_DIST = testlist
 ## for all programs that are just built from the single corresponding source
 ## file, we don't need per-target _SOURCES rules, automake will infer them
 ## correctly
-noinst_PROGRAMS = cfree ccreate1 manysplit userdup too_many_comms too_many_comms2 too_many_comms3 too_many_icomms too_many_icomms2
-
+noinst_PROGRAMS = cfree ccreate1 manysplit userdup too_many_comms too_many_comms2 too_many_comms3 too_many_icomms too_many_icomms2 \
```
Comment (raffenet) [inline test/mpi/errors/comm/Makefile.am]: @Bayyapu I thought we had agreed that new tests were going to be named `<function>_nullarg.c`. It looks like this version of the PR names them all `<function>_test.c`.

Comment (raffenet) [inline test/mpi/errors/comm/Makefile.am]: Oops, commented on the wrong PR.
